    > Version 1.0 of the attribute classification exercise

In [6]:
import mysql.connector
import pandas as pd
mydb=mysql.connector.connect(host='report-reader.cluster-custom-c8oe0gvszktr.ap-south-1.rds.amazonaws.com',
                             user='skreadonlyuser_api',
                             password='XUerT9JxnWnEstSA')
cursor=mydb.cursor()

In [7]:
cursor.execute('''
Select 
v.variantId as VariantId,
upper(clr.colorName) as Color,
clr.colorHex as ColorHex,
case when Upper(cm.name) = 'T-SHIRT' then 'T-SHIRTS' else cm.name end as Category,
cm1.name as MainCategory,
cm2.name as SuperCategory,
pa.attribute as ProductAttribute
from shoekonnect_live.variants as v 
INNER JOIN shoekonnect_live.colors as clr on clr.uniqueId = v.colorId
INNER JOIN shoekonnect_live.products as p on p.productid = v.productid 
INNER JOIN shoekonnect_live.category_master as cm on cm.categoryid = p.subcategoryid
INNER JOIN shoekonnect_live.category_master as cm1 on cm1.categoryid = cm.parentid
INNER JOIN shoekonnect_live.category_master as cm2 on cm2.categoryid = cm1.parentid
INNER JOIN shoekonnect_live.products_attribute as pa on pa.productid = p.productid
where upper(cm.name) IN ('JEANS', 'T-SHIRTS', 'T-SHIRT') and pa.attribute IS NOT NULL
''')

In [8]:
import pandas as pd
import json
rows=cursor.fetchall()
columns=[i[0] for i in  cursor.description]
df=pd.DataFrame(rows,columns=columns)
df

VariantId  Category  MainCategory SuperCategory  \
0        2089751     Jeans           Men      Apparels   
1        2090449     Jeans           Men      Apparels   
2        2090496     Jeans           Men      Apparels   
3        2090530     Jeans           Men      Apparels   
4        2092156     Jeans           Men      Apparels   
...          ...       ...           ...           ...   
59155    2642113  t-shirts  Girls - Kids      Apparels   
59156    2644147  t-shirts  Girls - Kids      Apparels   
59157    2648433  t-shirts  Girls - Kids      Apparels   
59158    2648434  t-shirts  Girls - Kids      Apparels   
59159    2648436  t-shirts  Girls - Kids      Apparels   

                                        ProductAttribute  
0      [{"name": "Fabric", "value": "Poly Knitted Lyc...  
1      [{"name": "Fabric", "value": "Poly Knitted Lyc...  
2      [{"name": "Fabric", "value": "Cotton By Cotton...  
3      [{"name": "Fabric", "value": "Cotton By Cotton...  
4      [{"name": "Fabric", "value": "Poly Knitted 12-...  
...                                                  ...  
59155  [{"name": "fabric", "value": "Cotton Lycra Bio...  
59156  [{"name": "neck stiching", "value": "overlock ...  
59157  [{"name": "print type", "value": "printed", "v...  
59158  [{"name": "print type", "value": "printed", "v...  
59159  [{"name": "print type", "value": "printed", "v...  

[59160 rows x 5 columns]

In [9]:
df

VariantId  Category  MainCategory SuperCategory  \
0        2089751     Jeans           Men      Apparels   
1        2090449     Jeans           Men      Apparels   
2        2090496     Jeans           Men      Apparels   
3        2090530     Jeans           Men      Apparels   
4        2092156     Jeans           Men      Apparels   
...          ...       ...           ...           ...   
59155    2642113  t-shirts  Girls - Kids      Apparels   
59156    2644147  t-shirts  Girls - Kids      Apparels   
59157    2648433  t-shirts  Girls - Kids      Apparels   
59158    2648434  t-shirts  Girls - Kids      Apparels   
59159    2648436  t-shirts  Girls - Kids      Apparels   

                                        ProductAttribute  
0      [{"name": "Fabric", "value": "Poly Knitted Lyc...  
1      [{"name": "Fabric", "value": "Poly Knitted Lyc...  
2      [{"name": "Fabric", "value": "Cotton By Cotton...  
3      [{"name": "Fabric", "value": "Cotton By Cotton...  
4      [{"name": "Fabric", "value": "Poly Knitted 12-...  
...                                                  ...  
59155  [{"name": "fabric", "value": "Cotton Lycra Bio...  
59156  [{"name": "neck stiching", "value": "overlock ...  
59157  [{"name": "print type", "value": "printed", "v...  
59158  [{"name": "print type", "value": "printed", "v...  
59159  [{"name": "print type", "value": "printed", "v...  

[59160 rows x 5 columns]

In [10]:
# Function to process each chunk of data
def process_data(df):
    # Parse the JSON in 'ProductAttribute'
    df['ProductAttribute'] = df['ProductAttribute'].apply(json.loads)

    # Explode the 'ProductAttribute' column so each list element becomes its own row
    df_exploded = df.explode('ProductAttribute')

    # Normalize the JSON structure into a dataframe
    df_normalized = pd.json_normalize(df_exploded['ProductAttribute'])

    # Concatenate with the original 'VariantId' column
    df_final = pd.concat([df_exploded['VariantId'].reset_index(drop=True), df_normalized[['name', 'value']]], axis=1)

    # Pivot the table to get 'name' as columns and 'value' as data points
    df_pivot = df_final.pivot_table(index='VariantId', columns='name', values='value', aggfunc='first').reset_index()

    return df_pivot

# Process the data
df_result = process_data(df)

# Show the result
df_result

name   VariantId Belt Type Buttons Closure Design     Fabric  \
0        1513399       NaN     NaN     NaN    NaN     Cotton   
1        1530083       NaN     NaN     NaN    NaN  Polyester   
2        1547548       NaN     NaN     NaN    NaN      Denim   
3        1547567       NaN     NaN     NaN    NaN      Denim   
4        1569064       NaN     NaN     NaN    NaN    Hosiery   
...          ...       ...     ...     ...    ...        ...   
59155    2652385       NaN     NaN     NaN    NaN        NaN   
59156    2652425       NaN     NaN     NaN    NaN        NaN   
59157    2652426       NaN     NaN     NaN    NaN        NaN   
59158    2652427       NaN     NaN     NaN    NaN        NaN   
59159    2652442       NaN     NaN     NaN    NaN        NaN   

name  Fabric Composition Fabric Grade Fabric Weight          Fit  ...  \
0                    NaN      200 GSM           NaN          NaN  ...   
1                    NaN          NaN           NaN          NaN  ...   
2                    NaN          NaN           NaN  Regular Fit  ...   
3                    NaN          NaN           NaN  Regular Fit  ...   
4                    NaN          NaN       180 GSM          NaN  ...   
...                  ...          ...           ...          ...  ...   
59155                NaN          NaN           NaN          NaN  ...   
59156                NaN          NaN           NaN          NaN  ...   
59157                NaN          NaN           NaN          NaN  ...   
59158                NaN          NaN           NaN          NaN  ...   
59159                NaN          NaN           NaN          NaN  ...   

name    softner stiching quality strechability      stretch style top fabric  \
0           NaN              NaN           NaN          NaN   NaN        NaN   
1           NaN              NaN           NaN          NaN   NaN        NaN   
2           NaN              NaN           NaN          NaN   NaN        NaN   
3           NaN              NaN           NaN          NaN   NaN        NaN   
4           NaN              NaN           NaN          NaN   NaN        NaN   
...         ...              ...           ...          ...   ...        ...   
59155        no              NaN           NaN           no   NaN        NaN   
59156  softener              NaN           NaN  stretchable   NaN        NaN   
59157  softener              NaN           NaN  stretchable   NaN        NaN   
59158  softener              NaN           NaN  stretchable   NaN        NaN   
59159       NaN              NaN           NaN          NaN   NaN        NaN   

name    waist rise wash effect wash quality work  
0              NaN         NaN          NaN  NaN  
1              NaN         NaN          NaN  NaN  
2              NaN         NaN          NaN  NaN  
3              NaN         NaN          NaN  NaN  
4              NaN         NaN          NaN  NaN  
...            ...         ...          ...  ...  
59155  medium rise       faded     bio wash  NaN  
59156  medium rise       faded     bio wash  NaN  
59157  medium rise       faded     bio wash  NaN  
59158  medium rise       faded     bio wash  NaN  
59159          NaN         NaN          NaN  NaN  

[59160 rows x 98 columns]

In [11]:
# df_result = df_result.fillna("NaN")
# df_result

In [12]:
## Step for data transformation for final usage
key_columns = ['name','VariantId']
for col in df_result.columns:
    if col not in key_columns:
        df_result[col] = df_result.apply(lambda row: f"{{{col}: {row[col]}}}" if pd.notnull(row[col]) else "", axis=1)
df_result

name   VariantId Belt Type Buttons Closure Design               Fabric  \
0        1513399                                      {Fabric: Cotton}   
1        1530083                                   {Fabric: Polyester}   
2        1547548                                       {Fabric: Denim}   
3        1547567                                       {Fabric: Denim}   
4        1569064                                     {Fabric: Hosiery}   
...          ...       ...     ...     ...    ...                  ...   
59155    2652385                                                         
59156    2652425                                                         
59157    2652426                                                         
59158    2652427                                                         
59159    2652442                                                         

name  Fabric Composition             Fabric Grade             Fabric Weight  \
0                         {Fabric Grade: 200 GSM}                             
1                                                                             
2                                                                             
3                                                                             
4                                                  {Fabric Weight: 180 GSM}   
...                  ...                      ...                       ...   
59155                                                                         
59156                                                                         
59157                                                                         
59158                                                                         
59159                                                                         

name                  Fit  ...              softner stiching quality  \
0                          ...                                         
1                          ...                                         
2      {Fit: Regular Fit}  ...                                         
3      {Fit: Regular Fit}  ...                                         
4                          ...                                         
...                   ...  ...                  ...              ...   
59155                      ...        {softner: no}                    
59156                      ...  {softner: softener}                    
59157                      ...  {softner: softener}                    
59158                      ...  {softner: softener}                    
59159                      ...                                         

name  strechability                 stretch style top fabric  \
0                                                              
1                                                              
2                                                              
3                                                              
4                                                              
...             ...                     ...   ...        ...   
59155                         {stretch: no}                    
59156                {stretch: stretchable}                    
59157                {stretch: stretchable}                    
59158                {stretch: stretchable}                    
59159                                                          

name                  waist rise           wash effect  \
0                                                        
1                                                        
2                                                        
3                                                        
4                                                        
...                          ...                   ...   
59155  {waist rise: medium rise}  {wash effect: faded}   
59156  {waist rise: medium rise}  {wash effect: faded}   
59157 

In [13]:
variant_ids = df_result[['VariantId']]
attributes = df_result.drop('VariantId', axis = 1)
encoded_attributes = pd.get_dummies(attributes)
encoded_attributes

Belt Type_  Belt Type_{Belt Type: Elastic}  Buttons_  \
0            True                           False      True   
1            True                           False      True   
2            True                           False      True   
3            True                           False      True   
4            True                           False      True   
...           ...                             ...       ...   
59155        True                           False      True   
59156        True                           False      True   
59157        True                           False      True   
59158        True                           False      True   
59159        True                           False      True   

       Buttons_{Buttons: 10 oz}  Buttons_{Buttons: 11 oz}  \
0                         False                     False   
1                         False                     False   
2                         False                     False   
3                         False                     False   
4                         False                     False   
...                         ...                       ...   
59155                     False                     False   
59156                     False                     False   
59157                     False                     False   
59158                     False                     False   
59159                     False                     False   

       Buttons_{Buttons: 2 Buttons}  Buttons_{Buttons: 3 Buttons}  \
0                             False                         False   
1                             False                         False   
2                             False                         False   
3                             False                         False   
4                             False                         False   
...                             ...                           ...   
59155                         False                         False   
59156                         False                         False   
59157                         False                         False   
59158                         False                         False   
59159                         False                         False   

       Buttons_{Buttons: 4 or More Buttons}  Buttons_{Buttons: 5 Buttons}  \
0                                     False                         False   
1                                     False                         False   
2                                     False                         False   
3                                     False                         False   
4                                     False                         False   
...                                     ...                           ...   
59155                                 False                         False   
59156                                 False                         False   
59157                                 False                         False   
59158                                 False                         False   
59159                                 False                         False   

       Buttons_{Buttons: N/A}  ...  wash quality_{wash quality: bio wash}  \
0                       False  ...                                  False   
1                       False  ...                                  False   
2                       False  ...                                  False   
3                       False  ...                                  False   
4                       False  ...                                  False   
...                       ...  ...                                    ...   
59155                   False  ...                                   True   
59156                   False  ...                                   True   
59157                   False  ...                                   T

In [14]:
encoded_attributes[encoded_attributes.select_dtypes(include=['bool']).columns] = encoded_attributes.select_dtypes(include=['bool']).astype(int)

In [15]:
encoded_attributes

Belt Type_  Belt Type_{Belt Type: Elastic}  Buttons_  \
0               1                               0         1   
1               1                               0         1   
2               1                               0         1   
3               1                               0         1   
4               1                               0         1   
...           ...                             ...       ...   
59155           1                               0         1   
59156           1                               0         1   
59157           1                               0         1   
59158           1                               0         1   
59159           1                               0         1   

       Buttons_{Buttons: 10 oz}  Buttons_{Buttons: 11 oz}  \
0                             0                         0   
1                             0                         0   
2                             0                         0   
3                             0                         0   
4                             0                         0   
...                         ...                       ...   
59155                         0                         0   
59156                         0                         0   
59157                         0                         0   
59158                         0                         0   
59159                         0                         0   

       Buttons_{Buttons: 2 Buttons}  Buttons_{Buttons: 3 Buttons}  \
0                                 0                             0   
1                                 0                             0   
2                                 0                             0   
3                                 0                             0   
4                                 0                             0   
...                             ...                           ...   
59155                             0                             0   
59156                             0                             0   
59157                             0                             0   
59158                             0                             0   
59159                             0                             0   

       Buttons_{Buttons: 4 or More Buttons}  Buttons_{Buttons: 5 Buttons}  \
0                                         0                             0   
1                                         0                             0   
2                                         0                             0   
3                                         0                             0   
4                                         0                             0   
...                                     ...                           ...   
59155                                     0                             0   
59156                                     0                             0   
59157                                     0                             0   
59158                                     0                             0   
59159                                     0                             0   

       Buttons_{Buttons: N/A}  ...  wash quality_{wash quality: bio wash}  \
0                           0  ...                                      0   
1                           0  ...                                      0   
2                           0  ...                                      0   
3                           0  ...                                      0   
4                           0  ...                                      0   
...                       ...  ...                                    ...   
59155                       0  ...                                      1   
59156                       0  ...                                      1   
59157                       0  ...                                    

In [16]:
df_result = pd.concat([variant_ids, encoded_attributes], axis=1)
df_result

VariantId  Belt Type_  Belt Type_{Belt Type: Elastic}  Buttons_  \
0        1513399           1                               0         1   
1        1530083           1                               0         1   
2        1547548           1                               0         1   
3        1547567           1                               0         1   
4        1569064           1                               0         1   
...          ...         ...                             ...       ...   
59155    2652385           1                               0         1   
59156    2652425           1                               0         1   
59157    2652426           1                               0         1   
59158    2652427           1                               0         1   
59159    2652442           1                               0         1   

       Buttons_{Buttons: 10 oz}  Buttons_{Buttons: 11 oz}  \
0                             0                         0   
1                             0                         0   
2                             0                         0   
3                             0                         0   
4                             0                         0   
...                         ...                       ...   
59155                         0                         0   
59156                         0                         0   
59157                         0                         0   
59158                         0                         0   
59159                         0                         0   

       Buttons_{Buttons: 2 Buttons}  Buttons_{Buttons: 3 Buttons}  \
0                                 0                             0   
1                                 0                             0   
2                                 0                             0   
3                                 0                             0   
4                                 0                             0   
...                             ...                           ...   
59155                             0                             0   
59156                             0                             0   
59157                             0                             0   
59158                             0                             0   
59159                             0                             0   

       Buttons_{Buttons: 4 or More Buttons}  Buttons_{Buttons: 5 Buttons}  \
0                                         0                             0   
1                                         0                             0   
2                                         0                             0   
3                                         0                             0   
4                                         0                             0   
...                                     ...                           ...   
59155                                     0                             0   
59156                                     0                             0   
59157                                     0                             0   
59158                                     0                             0   
59159                                     0                             0   

       ...  wash quality_{wash quality: bio wash}  \
0      ...                                      0   
1      ...                                      0   
2      ...                                      0   
3      ...                                      0   
4      ...                                      0   
...    ...                                    ...   
59155  ...                                      1   
59156  ...                                      1   
59157  ...                                      1   
59158  ...                                      1   
59159  ...                                      

In [17]:
df_final_new = pd.merge(df,df_result,on='VariantId',how='left')
df_final_new

VariantId  Category  MainCategory SuperCategory  \
0        2089751     Jeans           Men      Apparels   
1        2090449     Jeans           Men      Apparels   
2        2090496     Jeans           Men      Apparels   
3        2090530     Jeans           Men      Apparels   
4        2092156     Jeans           Men      Apparels   
...          ...       ...           ...           ...   
59155    2642113  t-shirts  Girls - Kids      Apparels   
59156    2644147  t-shirts  Girls - Kids      Apparels   
59157    2648433  t-shirts  Girls - Kids      Apparels   
59158    2648434  t-shirts  Girls - Kids      Apparels   
59159    2648436  t-shirts  Girls - Kids      Apparels   

                                        ProductAttribute  Belt Type_  \
0      [{'name': 'Fabric', 'value': 'Poly Knitted Lyc...           1   
1      [{'name': 'Fabric', 'value': 'Poly Knitted Lyc...           1   
2      [{'name': 'Fabric', 'value': 'Cotton By Cotton...           1   
3      [{'name': 'Fabric', 'value': 'Cotton By Cotton...           1   
4      [{'name': 'Fabric', 'value': 'Poly Knitted 12-...           1   
...                                                  ...         ...   
59155  [{'name': 'fabric', 'value': 'Cotton Lycra Bio...           1   
59156  [{'name': 'neck stiching', 'value': 'overlock ...           1   
59157  [{'name': 'print type', 'value': 'printed', 'v...           1   
59158  [{'name': 'print type', 'value': 'printed', 'v...           1   
59159  [{'name': 'print type', 'value': 'printed', 'v...           1   

       Belt Type_{Belt Type: Elastic}  Buttons_  Buttons_{Buttons: 10 oz}  \
0                                   0         1                         0   
1                                   0         1                         0   
2                                   0         1                         0   
3                                   0         1                         0   
4                                   0         1                         0   
...                               ...       ...                       ...   
59155                               0         1                         0   
59156                               0         1                         0   
59157                               0         1                         0   
59158                               0         1                         0   
59159                               0         1                         0   

       Buttons_{Buttons: 11 oz}  ...  wash quality_{wash quality: bio wash}  \
0                             0  ...                                      0   
1                             0  ...                                      0   
2                             0  ...                                      0   
3                             0  ...                                      0   
4                             0  ...                                      0   
...                         ...  ...                                    ...   
59155                         0  ...                                      0   
59156                         0  ...                                      0   
59157                         0  ...                                      0   
59158                         0  ...                                      0   
59159                         0  ...                                      0   

       wash quality_{wash quality: enzyme wash}  \
0                                             0   
1                                             0   
2                                             0   
3                                             0   
4                                             0   
...                                         ...   
59155                                         0   
59156                                         0   
59157                                         0   
59158                                         0   
59159     

In [18]:
variants_list = df_final_new['VariantId'].to_list()

In [19]:
import pandas as pd
from pymongo import MongoClient
import concurrent.futures
from tqdm import tqdm

# MongoDB connection setup
client = MongoClient("mongodb+srv://BijnisLiveMongoReader:XUerT9JxnWnEstSA@prod-mongo.bijnis.com/bijnis_production?tls=false&ssl=false")  # Replace with your connection string
db = client["product_service_production"]  # Replace with your database name
col = db["variant_image_vectors"]  # Replace with your collection name

# Define the query filter
query_filter = {"variantId": {'$in': variants_list}}

# Function to fetch documents in batches
def fetch_documents(skip, limit):
    return list(col.find(query_filter).skip(skip).limit(limit))

# Function to fetch all documents using threading with progress bar
def fetch_all_documents(total_records, batch_size):
    all_documents = []
    num_batches = total_records // batch_size + (1 if total_records % batch_size > 0 else 0)  # Calculate total batches
    
    with concurrent.futures.ThreadPoolExecutor() as executor:
        futures = {executor.submit(fetch_documents, skip, batch_size): skip for skip in range(0, total_records, batch_size)}
        
        # Use tqdm to show progress bar
        for future in tqdm(concurrent.futures.as_completed(futures), total=num_batches, desc="Fetching documents"):
            try:
                result = future.result()
                all_documents.extend(result)
            except Exception as e:
                print(f"Error fetching documents: {e}")
                
    return all_documents

# Fetching the total count of documents that match the query
total_records = col.count_documents(query_filter)

# Define the batch size
batch_size = 10000  # Adjust this according to your memory capacity

# Fetch all documents
all_docs = fetch_all_documents(total_records, batch_size)

# Convert the result to a DataFrame
df_mongo = pd.DataFrame(all_docs)

# Display the DataFrame
df_mongo 

Fetching documents: 100%|██████████| 25/25 [20:44<00:00, 49.80s/it]  


_id  variantId   imageId  \
0       66eae9becd570ba0a14b5019    2132859  17458872   
1       66eae9becd570ba0a14b501a    2132859  17458873   
2       66eae9becd570ba0a14b501b    2132859  17458874   
3       66eae9e059566945cabe3f8e    2132860  17458875   
4       66eae9e059566945cabe3f8f    2132860  17458876   
...                          ...        ...       ...   
247118  66fa1eb37da136a03fe8a0fa    2643430  26680256   
247119  66fa1ea97da136a03fe8a0f7    2643431  26680257   
247120  66fa1ea97da136a03fe8a0f8    2643431  26680258   
247121  66fa1ea4928eb8beb6d3b6f0    2643432  26680259   
247122  66fa1ea4928eb8beb6d3b6f1    2643432  26680260   

                                                 imageUrl  \
0       https://bijnis.s3.ap-south-1.amazonaws.com/pro...   
1       https://bijnis.s3.ap-south-1.amazonaws.com/pro...   
2       https://bijnis.s3.ap-south-1.amazonaws.com/pro...   
3       https://bijnis.s3.ap-south-1.amazonaws.com/pro...   
4       https://bijnis.s3.ap-south-1.amazonaws.com/pro...   
...                                                   ...   
247118  https://bijnis.s3.amazonaws.com/PRODUCTION/upl...   
247119  https://bijnis.s3.amazonaws.com/PRODUCTION/upl...   
247120  https://bijnis.s3.amazonaws.com/PRODUCTION/upl...   
247121  https://bijnis.s3.amazonaws.com/PRODUCTION/upl...   
247122  https://bijnis.s3.amazonaws.com/PRODUCTION/upl...   

                                                   vector            createdOn  
0       [0.0, 0.23605331778526306, 0.0, 0.541036784648...  2024-09-18 20:24:54  
1       [0.0, 0.0, 0.00231779832392931, 0.231860458850...  2024-09-18 20:24:54  
2       [0.0, 1.1161534786224365, 0.11599060893058777,...  2024-09-18 20:24:54  
3       [0.3553023040294647, 0.06674028187990189, 0.0,...  2024-09-18 20:25:28  
4       [0.39543190598487854, 0.23390108346939087, 0.0...  2024-09-18 20:25:28  
...                                                   ...                  ...  
247118  [0.0, 0.0, 0.11995667964220047, 0.197625294327...  2024-09-30 09:14:51  
247119  [0.12077472358942032, 0.0, 0.04706697165966034...  2024-09-30 09:14:41  
247120  [0.0, 0.0, 0.5308622121810913, 0.0097048813477...  2024-09-30 09:14:41  
247121  [0.015275663696229458, 0.0, 0.0472753793001174...  2024-09-30 09:14:36  
247122  [0.0, 0.0, 0.015623705461621284, 0.01281377207...  2024-09-30 09:14:36  

[247123 rows x 6 columns]

In [20]:
df_mongo = df_mongo.rename(columns={"variantId":"VariantId"})
df_mongo

_id  VariantId   imageId  \
0       66eae9becd570ba0a14b5019    2132859  17458872   
1       66eae9becd570ba0a14b501a    2132859  17458873   
2       66eae9becd570ba0a14b501b    2132859  17458874   
3       66eae9e059566945cabe3f8e    2132860  17458875   
4       66eae9e059566945cabe3f8f    2132860  17458876   
...                          ...        ...       ...   
247118  66fa1eb37da136a03fe8a0fa    2643430  26680256   
247119  66fa1ea97da136a03fe8a0f7    2643431  26680257   
247120  66fa1ea97da136a03fe8a0f8    2643431  26680258   
247121  66fa1ea4928eb8beb6d3b6f0    2643432  26680259   
247122  66fa1ea4928eb8beb6d3b6f1    2643432  26680260   

                                                 imageUrl  \
0       https://bijnis.s3.ap-south-1.amazonaws.com/pro...   
1       https://bijnis.s3.ap-south-1.amazonaws.com/pro...   
2       https://bijnis.s3.ap-south-1.amazonaws.com/pro...   
3       https://bijnis.s3.ap-south-1.amazonaws.com/pro...   
4       https://bijnis.s3.ap-south-1.amazonaws.com/pro...   
...                                                   ...   
247118  https://bijnis.s3.amazonaws.com/PRODUCTION/upl...   
247119  https://bijnis.s3.amazonaws.com/PRODUCTION/upl...   
247120  https://bijnis.s3.amazonaws.com/PRODUCTION/upl...   
247121  https://bijnis.s3.amazonaws.com/PRODUCTION/upl...   
247122  https://bijnis.s3.amazonaws.com/PRODUCTION/upl...   

                                                   vector            createdOn  
0       [0.0, 0.23605331778526306, 0.0, 0.541036784648...  2024-09-18 20:24:54  
1       [0.0, 0.0, 0.00231779832392931, 0.231860458850...  2024-09-18 20:24:54  
2       [0.0, 1.1161534786224365, 0.11599060893058777,...  2024-09-18 20:24:54  
3       [0.3553023040294647, 0.06674028187990189, 0.0,...  2024-09-18 20:25:28  
4       [0.39543190598487854, 0.23390108346939087, 0.0...  2024-09-18 20:25:28  
...                                                   ...                  ...  
247118  [0.0, 0.0, 0.11995667964220047, 0.197625294327...  2024-09-30 09:14:51  
247119  [0.12077472358942032, 0.0, 0.04706697165966034...  2024-09-30 09:14:41  
247120  [0.0, 0.0, 0.5308622121810913, 0.0097048813477...  2024-09-30 09:14:41  
247121  [0.015275663696229458, 0.0, 0.0472753793001174...  2024-09-30 09:14:36  
247122  [0.0, 0.0, 0.015623705461621284, 0.01281377207...  2024-09-30 09:14:36  

[247123 rows x 6 columns]

In [21]:
df_flattened = pd.merge(df_mongo, df_final_new, on = 'VariantId', how='left')

In [22]:
df_flattened

_id  VariantId   imageId  \
0       66eae9becd570ba0a14b5019    2132859  17458872   
1       66eae9becd570ba0a14b501a    2132859  17458873   
2       66eae9becd570ba0a14b501b    2132859  17458874   
3       66eae9e059566945cabe3f8e    2132860  17458875   
4       66eae9e059566945cabe3f8f    2132860  17458876   
...                          ...        ...       ...   
247118  66fa1eb37da136a03fe8a0fa    2643430  26680256   
247119  66fa1ea97da136a03fe8a0f7    2643431  26680257   
247120  66fa1ea97da136a03fe8a0f8    2643431  26680258   
247121  66fa1ea4928eb8beb6d3b6f0    2643432  26680259   
247122  66fa1ea4928eb8beb6d3b6f1    2643432  26680260   

                                                 imageUrl  \
0       https://bijnis.s3.ap-south-1.amazonaws.com/pro...   
1       https://bijnis.s3.ap-south-1.amazonaws.com/pro...   
2       https://bijnis.s3.ap-south-1.amazonaws.com/pro...   
3       https://bijnis.s3.ap-south-1.amazonaws.com/pro...   
4       https://bijnis.s3.ap-south-1.amazonaws.com/pro...   
...                                                   ...   
247118  https://bijnis.s3.amazonaws.com/PRODUCTION/upl...   
247119  https://bijnis.s3.amazonaws.com/PRODUCTION/upl...   
247120  https://bijnis.s3.amazonaws.com/PRODUCTION/upl...   
247121  https://bijnis.s3.amazonaws.com/PRODUCTION/upl...   
247122  https://bijnis.s3.amazonaws.com/PRODUCTION/upl...   

                                                   vector  \
0       [0.0, 0.23605331778526306, 0.0, 0.541036784648...   
1       [0.0, 0.0, 0.00231779832392931, 0.231860458850...   
2       [0.0, 1.1161534786224365, 0.11599060893058777,...   
3       [0.3553023040294647, 0.06674028187990189, 0.0,...   
4       [0.39543190598487854, 0.23390108346939087, 0.0...   
...                                                   ...   
247118  [0.0, 0.0, 0.11995667964220047, 0.197625294327...   
247119  [0.12077472358942032, 0.0, 0.04706697165966034...   
247120  [0.0, 0.0, 0.5308622121810913, 0.0097048813477...   
247121  [0.015275663696229458, 0.0, 0.0472753793001174...   
247122  [0.0, 0.0, 0.015623705461621284, 0.01281377207...   

                  createdOn  Category MainCategory SuperCategory  \
0       2024-09-18 20:24:54  T-SHIRTS          Men      Apparels   
1       2024-09-18 20:24:54  T-SHIRTS          Men      Apparels   
2       2024-09-18 20:24:54  T-SHIRTS          Men      Apparels   
3       2024-09-18 20:25:28  T-SHIRTS          Men      Apparels   
4       2024-09-18 20:25:28  T-SHIRTS          Men      Apparels   
...                     ...       ...          ...           ...   
247118  2024-09-30 09:14:51     jeans          Men      Apparels   
247119  2024-09-30 09:14:41     jeans          Men      Apparels   
247120  2024-09-30 09:14:41     jeans          Men      Apparels   
247121  2024-09-30 09:14:36     jeans          Men      Apparels   
247122  2024-09-30 09:14:36     jeans          Men      Apparels   

                                         ProductAttribute  ...  \
0       [{'name': 'Style', 'value': 'Basic', 'valueId'...  ...   
1       [{'name': 'Style', 'value': 'Basic', 'valueId'...  ...   
2       [{'name': 'Style', 'value': 'Basic', 'valueId'...  ...   
3       [{'name': 'Style', 'value': 'Basic', 'valueId'...  ...   
4       [{'name': 'Style', 'value': 'Basic', 'valueId'...  ...   
...                                                   ...  ...   
247118  [{'name': 'fabric', 'value': 'Cotton Slub Lycr...  ...   
247119  [{'name': 'fabric', 'value': 'Cotton Slub Lycr...  ...   
247120  [{'name': 'fabric', 'value': 'Cotton Slub Lycr...  ...   
247121  [{'name': 'waist rise', 'value': 'medium rise'...  ...   
247122  [{'name': 'waist rise', 'value': 'medium rise'...  ...   

        wash quality_{wash quality: bio wash}  \
0                                           0   
1                                           0   
2                                           0   
3                                           0   
4                      

In [23]:
## Transformation of the vector column to individual labels for training purpose
vector_df = pd.DataFrame(df_flattened['vector'].to_list())
vector_df.columns = [f"V{i+1}" for i in range (vector_df.shape[1])]
df_expanded = pd.concat([df_flattened,vector_df],axis=1)
df_expanded = df_expanded.fillna(0)
df_expanded

_id  VariantId   imageId  \
0       66eae9becd570ba0a14b5019    2132859  17458872   
1       66eae9becd570ba0a14b501a    2132859  17458873   
2       66eae9becd570ba0a14b501b    2132859  17458874   
3       66eae9e059566945cabe3f8e    2132860  17458875   
4       66eae9e059566945cabe3f8f    2132860  17458876   
...                          ...        ...       ...   
247118  66fa1eb37da136a03fe8a0fa    2643430  26680256   
247119  66fa1ea97da136a03fe8a0f7    2643431  26680257   
247120  66fa1ea97da136a03fe8a0f8    2643431  26680258   
247121  66fa1ea4928eb8beb6d3b6f0    2643432  26680259   
247122  66fa1ea4928eb8beb6d3b6f1    2643432  26680260   

                                                 imageUrl  \
0       https://bijnis.s3.ap-south-1.amazonaws.com/pro...   
1       https://bijnis.s3.ap-south-1.amazonaws.com/pro...   
2       https://bijnis.s3.ap-south-1.amazonaws.com/pro...   
3       https://bijnis.s3.ap-south-1.amazonaws.com/pro...   
4       https://bijnis.s3.ap-south-1.amazonaws.com/pro...   
...                                                   ...   
247118  https://bijnis.s3.amazonaws.com/PRODUCTION/upl...   
247119  https://bijnis.s3.amazonaws.com/PRODUCTION/upl...   
247120  https://bijnis.s3.amazonaws.com/PRODUCTION/upl...   
247121  https://bijnis.s3.amazonaws.com/PRODUCTION/upl...   
247122  https://bijnis.s3.amazonaws.com/PRODUCTION/upl...   

                                                   vector  \
0       [0.0, 0.23605331778526306, 0.0, 0.541036784648...   
1       [0.0, 0.0, 0.00231779832392931, 0.231860458850...   
2       [0.0, 1.1161534786224365, 0.11599060893058777,...   
3       [0.3553023040294647, 0.06674028187990189, 0.0,...   
4       [0.39543190598487854, 0.23390108346939087, 0.0...   
...                                                   ...   
247118  [0.0, 0.0, 0.11995667964220047, 0.197625294327...   
247119  [0.12077472358942032, 0.0, 0.04706697165966034...   
247120  [0.0, 0.0, 0.5308622121810913, 0.0097048813477...   
247121  [0.015275663696229458, 0.0, 0.0472753793001174...   
247122  [0.0, 0.0, 0.015623705461621284, 0.01281377207...   

                  createdOn  Category MainCategory SuperCategory  \
0       2024-09-18 20:24:54  T-SHIRTS          Men      Apparels   
1       2024-09-18 20:24:54  T-SHIRTS          Men      Apparels   
2       2024-09-18 20:24:54  T-SHIRTS          Men      Apparels   
3       2024-09-18 20:25:28  T-SHIRTS          Men      Apparels   
4       2024-09-18 20:25:28  T-SHIRTS          Men      Apparels   
...                     ...       ...          ...           ...   
247118  2024-09-30 09:14:51     jeans          Men      Apparels   
247119  2024-09-30 09:14:41     jeans          Men      Apparels   
247120  2024-09-30 09:14:41     jeans          Men      Apparels   
247121  2024-09-30 09:14:36     jeans          Men      Apparels   
247122  2024-09-30 09:14:36     jeans          Men      Apparels   

                                         ProductAttribute  ...     V1271  \
0       [{'name': 'Style', 'value': 'Basic', 'valueId'...  ...  0.000000   
1       [{'name': 'Style', 'value': 'Basic', 'valueId'...  ...  0.143531   
2       [{'name': 'Style', 'value': 'Basic', 'valueId'...  ...  0.000000   
3       [{'name': 'Style', 'value': 'Basic', 'valueId'...  ...  0.000000   
4       [{'name': 'Style', 'value': 'Basic', 'valueId'...  ...  0.000000   
...                                                   ...  ...       ...   
247118  [{'name': 'fabric', 'value': 'Cotton Slub Lycr...  ...  0.015857   
247119  [{'name': 'fabric', 'value': 'Cotton Slub Lycr...  ...  0.000000   
247120  [{'name': 'fabric', 'value': 'Cotton Slub Lycr...  ...  0.000000   
247121  [{'name': 'waist rise', 'value': 'medium rise'...  ...  0.000000   
247122  [{'name': 'waist rise', 'value': 'medium rise'...  ...  0.000000   

           V1272     V1273     V1274     V1275     V1276     V1277     V1278  \
0       0.437545  0.379377  0.057807  0.027454  0.047639  0.000000  

In [24]:
#df_expanded.to_csv('Classification_Stage_1.csv')

In [25]:
## creating version_1 of the dataframe for further operations
filtered_columns = [col for col in df_expanded.columns if col.startswith('V') and col[1:].isdigit()] + ['Category']
filtered_df = df_expanded[filtered_columns]
filtered_df

V1        V2        V3        V4        V5        V6        V7  \
0       0.000000  0.236053  0.000000  0.541037  0.035586  0.439141  0.639344   
1       0.000000  0.000000  0.002318  0.231860  0.589387  0.315818  0.658652   
2       0.000000  1.116153  0.115991  0.000000  0.000000  0.000000  0.000000   
3       0.355302  0.066740  0.000000  0.064922  0.379176  0.000000  0.770836   
4       0.395432  0.233901  0.091000  0.243147  0.531631  0.538100  0.507115   
...          ...       ...       ...       ...       ...       ...       ...   
247118  0.000000  0.000000  0.119957  0.197625  0.000000  0.000000  0.213593   
247119  0.120775  0.000000  0.047067  0.145077  0.000000  0.294627  0.264174   
247120  0.000000  0.000000  0.530862  0.009705  0.000000  0.000000  0.130015   
247121  0.015276  0.000000  0.047275  0.000000  0.020218  0.047031  0.177078   
247122  0.000000  0.000000  0.015624  0.012814  0.000000  0.000000  0.331728   

              V8        V9       V10  ...     V1272     V1273     V1274  \
0       0.346441  0.878183  0.185443  ...  0.437545  0.379377  0.057807   
1       0.214725  0.085433  0.019882  ...  0.856780  0.561768  0.000000   
2       0.493843  0.028399  0.244287  ...  0.023127  0.981050  0.000000   
3       0.612497  0.190065  0.097854  ...  0.806925  0.791816  0.000000   
4       1.156292  0.107904  0.120507  ...  0.327475  0.739500  0.000000   
...          ...       ...       ...  ...       ...       ...       ...   
247118  0.148386  0.214779  0.542384  ...  0.292787  0.000000  0.636535   
247119  0.000000  1.250417  0.036951  ...  1.123815  0.000000  0.097913   
247120  0.477553  0.000000  0.833972  ...  0.885310  0.000000  0.239653   
247121  0.000000  0.345184  0.066882  ...  0.314013  0.000000  0.023167   
247122  0.661369  0.129900  0.172631  ...  0.269158  0.000000  0.014829   

           V1275     V1276     V1277     V1278     V1279     V1280  Category  
0       0.027454  0.047639  0.000000  0.000000  0.076710  0.042522  T-SHIRTS  
1       0.000000  0.053496  0.000000  0.000000  1.641669  0.000000  T-SHIRTS  
2       1.307044  0.000000  0.000000  0.000000  0.088604  0.000000  T-SHIRTS  
3       0.028479  0.296662  0.000000  0.000000  0.531868  0.000000  T-SHIRTS  
4       0.323359  0.019973  0.000000  0.000000  0.275930  0.000000  T-SHIRTS  
...          ...       ...       ...       ...       ...       ...       ...  
247118  0.230335  0.000000  0.289987  1.239995  0.132540  0.270762     jeans  
247119  0.000000  0.000000  0.023646  0.102291  0.582427  0.024401     jeans  
247120  0.439827  0.004465  0.000000  0.048496  0.211974  0.088579     jeans  
247121  0.016230  0.000000  0.000000  0.000000  0.000000  0.147468     jeans  
247122  0.187695  0.000000  0.002154  0.486474  0.748830  0.622428     jeans  

[247123 rows x 1281 columns]

In [26]:
filtered_df['Category'] = filtered_df['Category'].str.upper()
# filtered_df['MainCategory'] = filtered_df['MainCategory'].str.upper()
# filtered_df['SuperCategory'] = filtered_df['SuperCategory'].str.upper()
filtered_df

C:\Users\Admin\AppData\Local\Temp\ipykernel_8000\3972431576.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  filtered_df['Category'] = filtered_df['Category'].str.upper()


V1        V2        V3        V4        V5        V6        V7  \
0       0.000000  0.236053  0.000000  0.541037  0.035586  0.439141  0.639344   
1       0.000000  0.000000  0.002318  0.231860  0.589387  0.315818  0.658652   
2       0.000000  1.116153  0.115991  0.000000  0.000000  0.000000  0.000000   
3       0.355302  0.066740  0.000000  0.064922  0.379176  0.000000  0.770836   
4       0.395432  0.233901  0.091000  0.243147  0.531631  0.538100  0.507115   
...          ...       ...       ...       ...       ...       ...       ...   
247118  0.000000  0.000000  0.119957  0.197625  0.000000  0.000000  0.213593   
247119  0.120775  0.000000  0.047067  0.145077  0.000000  0.294627  0.264174   
247120  0.000000  0.000000  0.530862  0.009705  0.000000  0.000000  0.130015   
247121  0.015276  0.000000  0.047275  0.000000  0.020218  0.047031  0.177078   
247122  0.000000  0.000000  0.015624  0.012814  0.000000  0.000000  0.331728   

              V8        V9       V10  ...     V1272     V1273     V1274  \
0       0.346441  0.878183  0.185443  ...  0.437545  0.379377  0.057807   
1       0.214725  0.085433  0.019882  ...  0.856780  0.561768  0.000000   
2       0.493843  0.028399  0.244287  ...  0.023127  0.981050  0.000000   
3       0.612497  0.190065  0.097854  ...  0.806925  0.791816  0.000000   
4       1.156292  0.107904  0.120507  ...  0.327475  0.739500  0.000000   
...          ...       ...       ...  ...       ...       ...       ...   
247118  0.148386  0.214779  0.542384  ...  0.292787  0.000000  0.636535   
247119  0.000000  1.250417  0.036951  ...  1.123815  0.000000  0.097913   
247120  0.477553  0.000000  0.833972  ...  0.885310  0.000000  0.239653   
247121  0.000000  0.345184  0.066882  ...  0.314013  0.000000  0.023167   
247122  0.661369  0.129900  0.172631  ...  0.269158  0.000000  0.014829   

           V1275     V1276     V1277     V1278     V1279     V1280  Category  
0       0.027454  0.047639  0.000000  0.000000  0.076710  0.042522  T-SHIRTS  
1       0.000000  0.053496  0.000000  0.000000  1.641669  0.000000  T-SHIRTS  
2       1.307044  0.000000  0.000000  0.000000  0.088604  0.000000  T-SHIRTS  
3       0.028479  0.296662  0.000000  0.000000  0.531868  0.000000  T-SHIRTS  
4       0.323359  0.019973  0.000000  0.000000  0.275930  0.000000  T-SHIRTS  
...          ...       ...       ...       ...       ...       ...       ...  
247118  0.230335  0.000000  0.289987  1.239995  0.132540  0.270762     JEANS  
247119  0.000000  0.000000  0.023646  0.102291  0.582427  0.024401     JEANS  
247120  0.439827  0.004465  0.000000  0.048496  0.211974  0.088579     JEANS  
247121  0.016230  0.000000  0.000000  0.000000  0.000000  0.147468     JEANS  
247122  0.187695  0.000000  0.002154  0.486474  0.748830  0.622428     JEANS  

[247123 rows x 1281 columns]

### Start of the classification code
    > First Attempt with RandomForestClassifier

In [27]:
filtered_df

V1        V2        V3        V4        V5        V6        V7  \
0       0.000000  0.236053  0.000000  0.541037  0.035586  0.439141  0.639344   
1       0.000000  0.000000  0.002318  0.231860  0.589387  0.315818  0.658652   
2       0.000000  1.116153  0.115991  0.000000  0.000000  0.000000  0.000000   
3       0.355302  0.066740  0.000000  0.064922  0.379176  0.000000  0.770836   
4       0.395432  0.233901  0.091000  0.243147  0.531631  0.538100  0.507115   
...          ...       ...       ...       ...       ...       ...       ...   
247118  0.000000  0.000000  0.119957  0.197625  0.000000  0.000000  0.213593   
247119  0.120775  0.000000  0.047067  0.145077  0.000000  0.294627  0.264174   
247120  0.000000  0.000000  0.530862  0.009705  0.000000  0.000000  0.130015   
247121  0.015276  0.000000  0.047275  0.000000  0.020218  0.047031  0.177078   
247122  0.000000  0.000000  0.015624  0.012814  0.000000  0.000000  0.331728   

              V8        V9       V10  ...     V1272     V1273     V1274  \
0       0.346441  0.878183  0.185443  ...  0.437545  0.379377  0.057807   
1       0.214725  0.085433  0.019882  ...  0.856780  0.561768  0.000000   
2       0.493843  0.028399  0.244287  ...  0.023127  0.981050  0.000000   
3       0.612497  0.190065  0.097854  ...  0.806925  0.791816  0.000000   
4       1.156292  0.107904  0.120507  ...  0.327475  0.739500  0.000000   
...          ...       ...       ...  ...       ...       ...       ...   
247118  0.148386  0.214779  0.542384  ...  0.292787  0.000000  0.636535   
247119  0.000000  1.250417  0.036951  ...  1.123815  0.000000  0.097913   
247120  0.477553  0.000000  0.833972  ...  0.885310  0.000000  0.239653   
247121  0.000000  0.345184  0.066882  ...  0.314013  0.000000  0.023167   
247122  0.661369  0.129900  0.172631  ...  0.269158  0.000000  0.014829   

           V1275     V1276     V1277     V1278     V1279     V1280  Category  
0       0.027454  0.047639  0.000000  0.000000  0.076710  0.042522  T-SHIRTS  
1       0.000000  0.053496  0.000000  0.000000  1.641669  0.000000  T-SHIRTS  
2       1.307044  0.000000  0.000000  0.000000  0.088604  0.000000  T-SHIRTS  
3       0.028479  0.296662  0.000000  0.000000  0.531868  0.000000  T-SHIRTS  
4       0.323359  0.019973  0.000000  0.000000  0.275930  0.000000  T-SHIRTS  
...          ...       ...       ...       ...       ...       ...       ...  
247118  0.230335  0.000000  0.289987  1.239995  0.132540  0.270762     JEANS  
247119  0.000000  0.000000  0.023646  0.102291  0.582427  0.024401     JEANS  
247120  0.439827  0.004465  0.000000  0.048496  0.211974  0.088579     JEANS  
247121  0.016230  0.000000  0.000000  0.000000  0.000000  0.147468     JEANS  
247122  0.187695  0.000000  0.002154  0.486474  0.748830  0.622428     JEANS  

[247123 rows x 1281 columns]

In [28]:
from sklearn.model_selection import train_test_split 
X = filtered_df.drop(['Category'], axis = 1)
y = filtered_df[['Category']]
X_train, X_test, y_train, y_test = train_test_split(X,y, test_size = 0.2, random_state=42)

In [29]:
from sklearn.ensemble import RandomForestClassifier
rf = RandomForestClassifier(random_state=42)
rf.fit(X_train, y_train)



c:\Users\Admin\miniconda3\envs\my_env\lib\site-packages\sklearn\base.py:1152: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  return fit_method(estimator, *args, **kwargs)


RandomForestClassifier(random_state=42)

In [30]:
from sklearn.metrics import accuracy_score, f1_score
# Predict on test data
y_pred = rf.predict(X_test)

# Evaluate the model
print("Accuracy:", accuracy_score(y_test, y_pred))
print("F1 Score:", f1_score(y_test, y_pred, average='weighted'))

Accuracy: 0.988265048052605
F1 Score: 0.9882578914454111


In [50]:

import os
import pandas as pd
import numpy as np
import tensorflow as tf
from tensorflow.keras.applications import MobileNetV2
from tensorflow.keras.preprocessing.image import img_to_array
from PIL import Image, UnidentifiedImageError
import requests
from io import BytesIO
import streamlit as st
import json
import datetime


def extract_feature_vector(image_path):
    # Load the pre-trained MobileNetV2 model
    base_model = MobileNetV2(weights='imagenet', include_top=False, pooling='avg')
    model = tf.keras.Model(inputs=base_model.input, outputs=base_model.output)

    # Open the image using PIL
    img = Image.open(image_path)

    if img.format != 'JPEG':
        # Convert image to RGB if it is not
        if img.mode != 'RGB':
            img = img.convert('RGB')
        
        img.save(image_path, format='JPEG')

    # Resize the image to the size expected by MobileNetV2 (224x224)
    img = img.resize((224, 224))
    # Convert the image to a numpy array and preprocess for the model
    img_array = np.array(img)
    img_array = tf.keras.applications.mobilenet_v2.preprocess_input(img_array)
    img_array = np.expand_dims(img_array, axis=0)  # Add batch dimension

    # Extract the feature vector using the model
    feature_vector = model.predict(img_array)

    return feature_vector

# Example usage
image_path = r"C:\Users\Admin\Attribute_Classification\Product_Attribute_Classification\Image_Vectorization\Image_10.jpeg"
feature_vector = extract_feature_vector(image_path)
image_vector_df = pd.DataFrame(feature_vector)
image_vector_df.columns = [f"V{i+1}" for i in range (image_vector_df.shape[1])]
image_vector_df




1/1 [==============================] - 1s 687ms/step


V1   V2        V3       V4   V5   V6        V7        V8   V9    V10  ...  \
0  0.0  0.0  0.094462  0.08731  0.0  0.0  0.003508  0.047143  0.0  0.003  ...   

   V1271     V1272  V1273  V1274     V1275     V1276  V1277     V1278  V1279  \
0    0.0  0.564352    0.0    0.0  0.462524  0.023598    0.0  0.496213    0.0   

   V1280  
0    0.0  

[1 rows x 1280 columns]

In [51]:
y_new = rf.predict(image_vector_df)
y_new

array(['JEANS'], dtype=object)

In [52]:
import pickle

In [53]:
with open('random_forest_model.pkl', 'wb') as f:
    pickle.dump(rf, f)

In [54]:
with open('random_forest_model.pkl', 'rb') as f:
    rf = pickle.load(f)

In [55]:

import os
import pandas as pd
import numpy as np
import tensorflow as tf
from tensorflow.keras.applications import MobileNetV2
from tensorflow.keras.preprocessing.image import img_to_array
from PIL import Image, UnidentifiedImageError
import requests
from io import BytesIO
import streamlit as st
import json
import datetime


def extract_feature_vector(image_path):
    # Load the pre-trained MobileNetV2 model
    base_model = MobileNetV2(weights='imagenet', include_top=False, pooling='avg')
    model = tf.keras.Model(inputs=base_model.input, outputs=base_model.output)

    # Open the image using PIL
    img = Image.open(image_path)

    if img.format != 'JPEG':
        # Convert image to RGB if it is not
        if img.mode != 'RGB':
            img = img.convert('RGB')
        
        img.save(image_path, format='JPEG')

    # Resize the image to the size expected by MobileNetV2 (224x224)
    img = img.resize((224, 224))
    # Convert the image to a numpy array and preprocess for the model
    img_array = np.array(img)
    img_array = tf.keras.applications.mobilenet_v2.preprocess_input(img_array)
    img_array = np.expand_dims(img_array, axis=0)  # Add batch dimension

    # Extract the feature vector using the model
    feature_vector = model.predict(img_array)

    return feature_vector

# Example usage
image_path = r"C:\Users\Admin\Attribute_Classification\Product_Attribute_Classification\Image_Vectorization\Image_11.jpeg"
feature_vector = extract_feature_vector(image_path)
image_vector_df = pd.DataFrame(feature_vector)
image_vector_df.columns = [f"V{i+1}" for i in range (image_vector_df.shape[1])]
image_vector_df




1/1 [==============================] - 1s 666ms/step


V1   V2   V3      V4   V5        V6        V7        V8        V9  V10  \
0  0.0  0.0  0.0  2.3025  0.0  0.030358  0.088443  0.034139  1.260753  0.0   

   ...  V1271     V1272  V1273     V1274  V1275  V1276     V1277     V1278  \
0  ...    0.0  0.114412    0.0  0.079145    0.0    0.0  0.678712  0.351788   

   V1279     V1280  
0    0.0  0.014979  

[1 rows x 1280 columns]

In [ ]:
y_new_2 = rf.predict(image_vector_df)
y_new_2

array(['JEANS'], dtype=object)

    > Trying SGDClassifier (Stochastic Gradient Descent) with incremental learning using chunks for faster and efficient execution

In [58]:
from sklearn.linear_model import SGDClassifier
sgd_clf = SGDClassifier(max_iter=1000, tol=1e-3, random_state=42)

# Scaling in chunks 
batch_size = 10000
for start in range(0, len(X_train),batch_size):
    end = start + batch_size
    X_batch = X_train[start:end]
    y_batch = y_train[start:end]
    
    sgd_clf.partial_fit(X_batch,y_batch, classes = np.unique(y))

y_pred_1 = sgd_clf.predict(X_test)

c:\Users\Admin\miniconda3\envs\my_env\lib\site-packages\sklearn\utils\validation.py:1183: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
c:\Users\Admin\miniconda3\envs\my_env\lib\site-packages\sklearn\utils\validation.py:1183: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
c:\Users\Admin\miniconda3\envs\my_env\lib\site-packages\sklearn\utils\validation.py:1183: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
c:\Users\Admin\miniconda3\envs\my_env\lib\site-packages\sklearn\utils\validation.py:1183: DataConversionWarning: A column-vector y was passed when a 1d array was e

In [70]:
# Accuracy Testing 
from sklearn.metrics import accuracy_score, f1_score
# Evaluate the model
print("Accuracy:", accuracy_score(y_test, y_pred_1))
print("F1 Score:", f1_score(y_test, y_pred_1, average='weighted'))

Accuracy: 0.984279210925645
F1 Score: 0.9842629074210562


    > SGDClassifier has significantly reduced the training time with marginal change (~ 0.50 %) in the accuracy

### Code for the Color Classification

    > Trying first time with the SGDClassifier for color classification and training

In [71]:
df_expanded

_id  VariantId   imageId  \
0       66eae9becd570ba0a14b5019    2132859  17458872   
1       66eae9becd570ba0a14b501a    2132859  17458873   
2       66eae9becd570ba0a14b501b    2132859  17458874   
3       66eae9e059566945cabe3f8e    2132860  17458875   
4       66eae9e059566945cabe3f8f    2132860  17458876   
...                          ...        ...       ...   
247118  66fa1eb37da136a03fe8a0fa    2643430  26680256   
247119  66fa1ea97da136a03fe8a0f7    2643431  26680257   
247120  66fa1ea97da136a03fe8a0f8    2643431  26680258   
247121  66fa1ea4928eb8beb6d3b6f0    2643432  26680259   
247122  66fa1ea4928eb8beb6d3b6f1    2643432  26680260   

                                                 imageUrl  \
0       https://bijnis.s3.ap-south-1.amazonaws.com/pro...   
1       https://bijnis.s3.ap-south-1.amazonaws.com/pro...   
2       https://bijnis.s3.ap-south-1.amazonaws.com/pro...   
3       https://bijnis.s3.ap-south-1.amazonaws.com/pro...   
4       https://bijnis.s3.ap-south-1.amazonaws.com/pro...   
...                                                   ...   
247118  https://bijnis.s3.amazonaws.com/PRODUCTION/upl...   
247119  https://bijnis.s3.amazonaws.com/PRODUCTION/upl...   
247120  https://bijnis.s3.amazonaws.com/PRODUCTION/upl...   
247121  https://bijnis.s3.amazonaws.com/PRODUCTION/upl...   
247122  https://bijnis.s3.amazonaws.com/PRODUCTION/upl...   

                                                   vector  \
0       [0.0, 0.23605331778526306, 0.0, 0.541036784648...   
1       [0.0, 0.0, 0.00231779832392931, 0.231860458850...   
2       [0.0, 1.1161534786224365, 0.11599060893058777,...   
3       [0.3553023040294647, 0.06674028187990189, 0.0,...   
4       [0.39543190598487854, 0.23390108346939087, 0.0...   
...                                                   ...   
247118  [0.0, 0.0, 0.11995667964220047, 0.197625294327...   
247119  [0.12077472358942032, 0.0, 0.04706697165966034...   
247120  [0.0, 0.0, 0.5308622121810913, 0.0097048813477...   
247121  [0.015275663696229458, 0.0, 0.0472753793001174...   
247122  [0.0, 0.0, 0.015623705461621284, 0.01281377207...   

                  createdOn  Category MainCategory SuperCategory  \
0       2024-09-18 20:24:54  T-SHIRTS          Men      Apparels   
1       2024-09-18 20:24:54  T-SHIRTS          Men      Apparels   
2       2024-09-18 20:24:54  T-SHIRTS          Men      Apparels   
3       2024-09-18 20:25:28  T-SHIRTS          Men      Apparels   
4       2024-09-18 20:25:28  T-SHIRTS          Men      Apparels   
...                     ...       ...          ...           ...   
247118  2024-09-30 09:14:51     jeans          Men      Apparels   
247119  2024-09-30 09:14:41     jeans          Men      Apparels   
247120  2024-09-30 09:14:41     jeans          Men      Apparels   
247121  2024-09-30 09:14:36     jeans          Men      Apparels   
247122  2024-09-30 09:14:36     jeans          Men      Apparels   

                                         ProductAttribute  ...     V1271  \
0       [{'name': 'Style', 'value': 'Basic', 'valueId'...  ...  0.000000   
1       [{'name': 'Style', 'value': 'Basic', 'valueId'...  ...  0.143531   
2       [{'name': 'Style', 'value': 'Basic', 'valueId'...  ...  0.000000   
3       [{'name': 'Style', 'value': 'Basic', 'valueId'...  ...  0.000000   
4       [{'name': 'Style', 'value': 'Basic', 'valueId'...  ...  0.000000   
...                                                   ...  ...       ...   
247118  [{'name': 'fabric', 'value': 'Cotton Slub Lycr...  ...  0.015857   
247119  [{'name': 'fabric', 'value': 'Cotton Slub Lycr...  ...  0.000000   
247120  [{'name': 'fabric', 'value': 'Cotton Slub Lycr...  ...  0.000000   
247121  [{'name': 'waist rise', 'value': 'medium rise'...  ...  0.000000   
247122  [{'name': 'waist rise', 'value': 'medium rise'...  ...  0.000000   

           V1272     V1273     V1274     V1275     V1276     V1277     V1278  \
0       0.437545  0.379377  0.057807  0.027454  0.047639  0.000000  

In [74]:
import mysql.connector
import pandas as pd
mydb=mysql.connector.connect(host='report-reader.cluster-custom-c8oe0gvszktr.ap-south-1.rds.amazonaws.com',
                             user='skreadonlyuser_api',
                             password='XUerT9JxnWnEstSA')
cursor=mydb.cursor()

In [77]:
cursor.execute('''
Select 
v.Variantid as VariantId,
clr.ColorName as ColorName,
clr.colorHex as ColorHex
from shoekonnect_live.variants  as v
INNER JOIN shoekonnect_live.colors as clr on clr.uniqueId = v.colorId
''')

In [78]:
import pandas as pd
import json
rows=cursor.fetchall()
columns=[i[0] for i in  cursor.description]
df_colors=pd.DataFrame(rows,columns=columns)
df_colors

VariantId   ColorName ColorHex
0          1000000      Coffee  #4a342e
1          1000001       Black  #333333
2          1000002      Coffee  #4a342e
3          1000003     Mustard  #e1b228
4          1000004      Coffee  #4a342e
...            ...         ...      ...
1632912    2652541  Multicolor  #999999
1632913    2652542  Multicolor  #999999
1632914    2652543  Multicolor  #999999
1632915    2652544        Pink  #ec81b3
1632916    2652545  Multicolor  #999999

[1632917 rows x 3 columns]

In [79]:
df_colors['ColorName'] = df_colors['ColorName'].str.upper()
df_colors

VariantId   ColorName ColorHex
0          1000000      COFFEE  #4a342e
1          1000001       BLACK  #333333
2          1000002      COFFEE  #4a342e
3          1000003     MUSTARD  #e1b228
4          1000004      COFFEE  #4a342e
...            ...         ...      ...
1632912    2652541  MULTICOLOR  #999999
1632913    2652542  MULTICOLOR  #999999
1632914    2652543  MULTICOLOR  #999999
1632915    2652544        PINK  #ec81b3
1632916    2652545  MULTICOLOR  #999999

[1632917 rows x 3 columns]

In [80]:
color_class = pd.merge(df_expanded,df_colors, on='VariantId',how='left')
color_class

_id  VariantId   imageId  \
0       66eae9becd570ba0a14b5019    2132859  17458872   
1       66eae9becd570ba0a14b501a    2132859  17458873   
2       66eae9becd570ba0a14b501b    2132859  17458874   
3       66eae9e059566945cabe3f8e    2132860  17458875   
4       66eae9e059566945cabe3f8f    2132860  17458876   
...                          ...        ...       ...   
247118  66fa1eb37da136a03fe8a0fa    2643430  26680256   
247119  66fa1ea97da136a03fe8a0f7    2643431  26680257   
247120  66fa1ea97da136a03fe8a0f8    2643431  26680258   
247121  66fa1ea4928eb8beb6d3b6f0    2643432  26680259   
247122  66fa1ea4928eb8beb6d3b6f1    2643432  26680260   

                                                 imageUrl  \
0       https://bijnis.s3.ap-south-1.amazonaws.com/pro...   
1       https://bijnis.s3.ap-south-1.amazonaws.com/pro...   
2       https://bijnis.s3.ap-south-1.amazonaws.com/pro...   
3       https://bijnis.s3.ap-south-1.amazonaws.com/pro...   
4       https://bijnis.s3.ap-south-1.amazonaws.com/pro...   
...                                                   ...   
247118  https://bijnis.s3.amazonaws.com/PRODUCTION/upl...   
247119  https://bijnis.s3.amazonaws.com/PRODUCTION/upl...   
247120  https://bijnis.s3.amazonaws.com/PRODUCTION/upl...   
247121  https://bijnis.s3.amazonaws.com/PRODUCTION/upl...   
247122  https://bijnis.s3.amazonaws.com/PRODUCTION/upl...   

                                                   vector  \
0       [0.0, 0.23605331778526306, 0.0, 0.541036784648...   
1       [0.0, 0.0, 0.00231779832392931, 0.231860458850...   
2       [0.0, 1.1161534786224365, 0.11599060893058777,...   
3       [0.3553023040294647, 0.06674028187990189, 0.0,...   
4       [0.39543190598487854, 0.23390108346939087, 0.0...   
...                                                   ...   
247118  [0.0, 0.0, 0.11995667964220047, 0.197625294327...   
247119  [0.12077472358942032, 0.0, 0.04706697165966034...   
247120  [0.0, 0.0, 0.5308622121810913, 0.0097048813477...   
247121  [0.015275663696229458, 0.0, 0.0472753793001174...   
247122  [0.0, 0.0, 0.015623705461621284, 0.01281377207...   

                  createdOn  Category MainCategory SuperCategory  \
0       2024-09-18 20:24:54  T-SHIRTS          Men      Apparels   
1       2024-09-18 20:24:54  T-SHIRTS          Men      Apparels   
2       2024-09-18 20:24:54  T-SHIRTS          Men      Apparels   
3       2024-09-18 20:25:28  T-SHIRTS          Men      Apparels   
4       2024-09-18 20:25:28  T-SHIRTS          Men      Apparels   
...                     ...       ...          ...           ...   
247118  2024-09-30 09:14:51     jeans          Men      Apparels   
247119  2024-09-30 09:14:41     jeans          Men      Apparels   
247120  2024-09-30 09:14:41     jeans          Men      Apparels   
247121  2024-09-30 09:14:36     jeans          Men      Apparels   
247122  2024-09-30 09:14:36     jeans          Men      Apparels   

                                         ProductAttribute  ...     V1273  \
0       [{'name': 'Style', 'value': 'Basic', 'valueId'...  ...  0.379377   
1       [{'name': 'Style', 'value': 'Basic', 'valueId'...  ...  0.561768   
2       [{'name': 'Style', 'value': 'Basic', 'valueId'...  ...  0.981050   
3       [{'name': 'Style', 'value': 'Basic', 'valueId'...  ...  0.791816   
4       [{'name': 'Style', 'value': 'Basic', 'valueId'...  ...  0.739500   
...                                                   ...  ...       ...   
247118  [{'name': 'fabric', 'value': 'Cotton Slub Lycr...  ...  0.000000   
247119  [{'name': 'fabric', 'value': 'Cotton Slub Lycr...  ...  0.000000   
247120  [{'name': 'fabric', 'value': 'Cotton Slub Lycr...  ...  0.000000   
247121  [{'name': 'waist rise', 'value': 'medium rise'...  ...  0.000000   
247122  [{'name': 'waist rise', 'value': 'medium rise'...  ...  0.000000   

           V1274     V1275     V1276     V1277     V1278     V1279     V1280  \
0       0.057807  0.027454  0.047639  0.000000  0.000000  0.076710  

In [81]:
filtered_columns = [col for col in color_class.columns if col.startswith('V') and col[1:].isdigit()] + ['ColorName']
filtered_color_class = color_class[filtered_columns]
filtered_color_class

V1        V2        V3        V4        V5        V6        V7  \
0       0.000000  0.236053  0.000000  0.541037  0.035586  0.439141  0.639344   
1       0.000000  0.000000  0.002318  0.231860  0.589387  0.315818  0.658652   
2       0.000000  1.116153  0.115991  0.000000  0.000000  0.000000  0.000000   
3       0.355302  0.066740  0.000000  0.064922  0.379176  0.000000  0.770836   
4       0.395432  0.233901  0.091000  0.243147  0.531631  0.538100  0.507115   
...          ...       ...       ...       ...       ...       ...       ...   
247118  0.000000  0.000000  0.119957  0.197625  0.000000  0.000000  0.213593   
247119  0.120775  0.000000  0.047067  0.145077  0.000000  0.294627  0.264174   
247120  0.000000  0.000000  0.530862  0.009705  0.000000  0.000000  0.130015   
247121  0.015276  0.000000  0.047275  0.000000  0.020218  0.047031  0.177078   
247122  0.000000  0.000000  0.015624  0.012814  0.000000  0.000000  0.331728   

              V8        V9       V10  ...     V1272     V1273     V1274  \
0       0.346441  0.878183  0.185443  ...  0.437545  0.379377  0.057807   
1       0.214725  0.085433  0.019882  ...  0.856780  0.561768  0.000000   
2       0.493843  0.028399  0.244287  ...  0.023127  0.981050  0.000000   
3       0.612497  0.190065  0.097854  ...  0.806925  0.791816  0.000000   
4       1.156292  0.107904  0.120507  ...  0.327475  0.739500  0.000000   
...          ...       ...       ...  ...       ...       ...       ...   
247118  0.148386  0.214779  0.542384  ...  0.292787  0.000000  0.636535   
247119  0.000000  1.250417  0.036951  ...  1.123815  0.000000  0.097913   
247120  0.477553  0.000000  0.833972  ...  0.885310  0.000000  0.239653   
247121  0.000000  0.345184  0.066882  ...  0.314013  0.000000  0.023167   
247122  0.661369  0.129900  0.172631  ...  0.269158  0.000000  0.014829   

           V1275     V1276     V1277     V1278     V1279     V1280   ColorName  
0       0.027454  0.047639  0.000000  0.000000  0.076710  0.042522       BLACK  
1       0.000000  0.053496  0.000000  0.000000  1.641669  0.000000       BLACK  
2       1.307044  0.000000  0.000000  0.000000  0.088604  0.000000       BLACK  
3       0.028479  0.296662  0.000000  0.000000  0.531868  0.000000       BLACK  
4       0.323359  0.019973  0.000000  0.000000  0.275930  0.000000       BLACK  
...          ...       ...       ...       ...       ...       ...         ...  
247118  0.230335  0.000000  0.289987  1.239995  0.132540  0.270762  MULTICOLOR  
247119  0.000000  0.000000  0.023646  0.102291  0.582427  0.024401  MULTICOLOR  
247120  0.439827  0.004465  0.000000  0.048496  0.211974  0.088579  MULTICOLOR  
247121  0.016230  0.000000  0.000000  0.000000  0.000000  0.147468  MULTICOLOR  
247122  0.187695  0.000000  0.002154  0.486474  0.748830  0.622428  MULTICOLOR  

[247123 rows x 1281 columns]

In [82]:
from sklearn.model_selection import train_test_split 
X = filtered_color_class.drop(['ColorName'], axis = 1)
y = filtered_color_class[['ColorName']]
X_train, X_test, y_train, y_test = train_test_split(X,y, test_size = 0.2, random_state=42)

In [83]:
from sklearn.linear_model import SGDClassifier
sgd_clf = SGDClassifier(max_iter=1000, tol=1e-3, random_state=42)

# Scaling in chunks 
batch_size = 10000
for start in range(0, len(X_train),batch_size):
    end = start + batch_size
    X_batch = X_train[start:end]
    y_batch = y_train[start:end]
    
    sgd_clf.partial_fit(X_batch,y_batch, classes = np.unique(y))

y_pred_1 = sgd_clf.predict(X_test)

c:\Users\Admin\miniconda3\envs\my_env\lib\site-packages\sklearn\utils\validation.py:1183: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
c:\Users\Admin\miniconda3\envs\my_env\lib\site-packages\sklearn\utils\validation.py:1183: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
c:\Users\Admin\miniconda3\envs\my_env\lib\site-packages\sklearn\utils\validation.py:1183: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
c:\Users\Admin\miniconda3\envs\my_env\lib\site-packages\sklearn\utils\validation.py:1183: DataConversionWarning: A column-vector y was passed when a 1d array was e

In [84]:
# Accuracy Testing 
from sklearn.metrics import accuracy_score, f1_score
# Evaluate the model
print("Accuracy:", accuracy_score(y_test, y_pred_1))
print("F1 Score:", f1_score(y_test, y_pred_1, average='weighted'))

Accuracy: 0.32507840161861407
F1 Score: 0.293523874946922


In [108]:

import os
import pandas as pd
import numpy as np
import tensorflow as tf
from tensorflow.keras.applications import MobileNetV2
from tensorflow.keras.preprocessing.image import img_to_array
from PIL import Image, UnidentifiedImageError
import requests
from io import BytesIO
import streamlit as st
import json
import datetime


def extract_feature_vector(image_path):
    # Load the pre-trained MobileNetV2 model
    base_model = MobileNetV2(weights='imagenet', include_top=False, pooling='avg')
    model = tf.keras.Model(inputs=base_model.input, outputs=base_model.output)

    # Open the image using PIL
    img = Image.open(image_path)

    if img.format != 'JPEG':
        # Convert image to RGB if it is not
        if img.mode != 'RGB':
            img = img.convert('RGB')
        
        img.save(image_path, format='JPEG')

    # Resize the image to the size expected by MobileNetV2 (224x224)
    img = img.resize((224, 224))
    # Convert the image to a numpy array and preprocess for the model
    img_array = np.array(img)
    img_array = tf.keras.applications.mobilenet_v2.preprocess_input(img_array)
    img_array = np.expand_dims(img_array, axis=0)  # Add batch dimension

    # Extract the feature vector using the model
    feature_vector = model.predict(img_array)

    return feature_vector

# Example usage
image_path = r"C:\Users\Admin\Attribute_Classification\Product_Attribute_Classification\Image_Vectorization\Image_16.jpeg"
feature_vector = extract_feature_vector(image_path)
image_vector_df = pd.DataFrame(feature_vector)
image_vector_df.columns = [f"V{i+1}" for i in range (image_vector_df.shape[1])]
image_vector_df 

1/1 [==============================] - 1s 800ms/step


V1        V2   V3   V4        V5   V6        V7   V8       V9  V10  \
0  0.031981  0.129628  0.0  0.0  0.683886  0.0  0.535589  0.0  0.01512  0.0   

   ...  V1271     V1272     V1273     V1274     V1275     V1276  V1277  \
0  ...    0.0  0.897217  0.887154  0.042489  0.060225  0.206238    0.0   

      V1278     V1279     V1280  
0  0.088989  0.016041  0.050159  

[1 rows x 1280 columns]

In [109]:
y_pred_1 = sgd_clf.predict(image_vector_df)
y_pred_1

array(['WHITE'], dtype='<U255')

### Code for the MainCategory Classification

    > Training for the Maincategory

In [113]:
color_class

_id  VariantId   imageId  \
0       66eae9becd570ba0a14b5019    2132859  17458872   
1       66eae9becd570ba0a14b501a    2132859  17458873   
2       66eae9becd570ba0a14b501b    2132859  17458874   
3       66eae9e059566945cabe3f8e    2132860  17458875   
4       66eae9e059566945cabe3f8f    2132860  17458876   
...                          ...        ...       ...   
247118  66fa1eb37da136a03fe8a0fa    2643430  26680256   
247119  66fa1ea97da136a03fe8a0f7    2643431  26680257   
247120  66fa1ea97da136a03fe8a0f8    2643431  26680258   
247121  66fa1ea4928eb8beb6d3b6f0    2643432  26680259   
247122  66fa1ea4928eb8beb6d3b6f1    2643432  26680260   

                                                 imageUrl  \
0       https://bijnis.s3.ap-south-1.amazonaws.com/pro...   
1       https://bijnis.s3.ap-south-1.amazonaws.com/pro...   
2       https://bijnis.s3.ap-south-1.amazonaws.com/pro...   
3       https://bijnis.s3.ap-south-1.amazonaws.com/pro...   
4       https://bijnis.s3.ap-south-1.amazonaws.com/pro...   
...                                                   ...   
247118  https://bijnis.s3.amazonaws.com/PRODUCTION/upl...   
247119  https://bijnis.s3.amazonaws.com/PRODUCTION/upl...   
247120  https://bijnis.s3.amazonaws.com/PRODUCTION/upl...   
247121  https://bijnis.s3.amazonaws.com/PRODUCTION/upl...   
247122  https://bijnis.s3.amazonaws.com/PRODUCTION/upl...   

                                                   vector  \
0       [0.0, 0.23605331778526306, 0.0, 0.541036784648...   
1       [0.0, 0.0, 0.00231779832392931, 0.231860458850...   
2       [0.0, 1.1161534786224365, 0.11599060893058777,...   
3       [0.3553023040294647, 0.06674028187990189, 0.0,...   
4       [0.39543190598487854, 0.23390108346939087, 0.0...   
...                                                   ...   
247118  [0.0, 0.0, 0.11995667964220047, 0.197625294327...   
247119  [0.12077472358942032, 0.0, 0.04706697165966034...   
247120  [0.0, 0.0, 0.5308622121810913, 0.0097048813477...   
247121  [0.015275663696229458, 0.0, 0.0472753793001174...   
247122  [0.0, 0.0, 0.015623705461621284, 0.01281377207...   

                  createdOn  Category MainCategory SuperCategory  \
0       2024-09-18 20:24:54  T-SHIRTS          Men      Apparels   
1       2024-09-18 20:24:54  T-SHIRTS          Men      Apparels   
2       2024-09-18 20:24:54  T-SHIRTS          Men      Apparels   
3       2024-09-18 20:25:28  T-SHIRTS          Men      Apparels   
4       2024-09-18 20:25:28  T-SHIRTS          Men      Apparels   
...                     ...       ...          ...           ...   
247118  2024-09-30 09:14:51     jeans          Men      Apparels   
247119  2024-09-30 09:14:41     jeans          Men      Apparels   
247120  2024-09-30 09:14:41     jeans          Men      Apparels   
247121  2024-09-30 09:14:36     jeans          Men      Apparels   
247122  2024-09-30 09:14:36     jeans          Men      Apparels   

                                         ProductAttribute  ...     V1273  \
0       [{'name': 'Style', 'value': 'Basic', 'valueId'...  ...  0.379377   
1       [{'name': 'Style', 'value': 'Basic', 'valueId'...  ...  0.561768   
2       [{'name': 'Style', 'value': 'Basic', 'valueId'...  ...  0.981050   
3       [{'name': 'Style', 'value': 'Basic', 'valueId'...  ...  0.791816   
4       [{'name': 'Style', 'value': 'Basic', 'valueId'...  ...  0.739500   
...                                                   ...  ...       ...   
247118  [{'name': 'fabric', 'value': 'Cotton Slub Lycr...  ...  0.000000   
247119  [{'name': 'fabric', 'value': 'Cotton Slub Lycr...  ...  0.000000   
247120  [{'name': 'fabric', 'value': 'Cotton Slub Lycr...  ...  0.000000   
247121  [{'name': 'waist rise', 'value': 'medium rise'...  ...  0.000000   
247122  [{'name': 'waist rise', 'value': 'medium rise'...  ...  0.000000   

           V1274     V1275     V1276     V1277     V1278     V1279     V1280  \
0       0.057807  0.027454  0.047639  0.000000  0.000000  0.076710  

In [115]:
filtered_columns = [col for col in color_class.columns if col.startswith('V') and col[1:].isdigit()] + ['ColorName']
filtered_color_class = color_class[filtered_columns]
filtered_color_class

MemoryError: Unable to allocate 2.36 GiB for an array with shape (1280, 247123) and data type float64

In [111]:
from sklearn.model_selection import train_test_split 
X = filtered_color_class.drop(['MainCategory'], axis = 1)
y = filtered_color_class[['MainCategory']]
X_train, X_test, y_train, y_test = train_test_split(X,y, test_size = 0.2, random_state=42)

MemoryError: Unable to allocate 2.36 GiB for an array with shape (1280, 247123) and data type float64

In [ ]:
from sklearn.linear_model import SGDClassifier
sgd_clf = SGDClassifier(max_iter=1000, tol=1e-3, random_state=42)

# Scaling in chunks 
batch_size = 10000
for start in range(0, len(X_train),batch_size):
    end = start + batch_size
    X_batch = X_train[start:end]
    y_batch = y_train[start:end]
    
    sgd_clf.partial_fit(X_batch,y_batch, classes = np.unique(y))

y_pred_1 = sgd_clf.predict(X_test)

In [ ]:
# Accuracy Testing 
from sklearn.metrics import accuracy_score, f1_score
# Evaluate the model
print("Accuracy:", accuracy_score(y_test, y_pred_1))
print("F1 Score:", f1_score(y_test, y_pred_1, average='weighted'))

#### Guidelines to achieve the desired outcome
+ Pull the desired data from the MongoDB and MySQL DataBase 

        > Fetching Data from MongoDB is a blocker in training the model for multiple subcategories as it taking huge time due to no mapping of the SubCategories/MainCategories/SuperCategories 
        

+ Create a flat file with all the parameters and product attributes for further processing
+ Encode the whole data for the parameters to get the binary values of the output labels
+ Transform the vectors for 1280 (size of the output vector from MobileNet_V2) labels in sequential columns as an input
+ Identify all the relevant parameters/labels required to be identified using the input vectors 

    > P_2 Priority
    
+ Split the whole data into training, testing and validation set in (70-20-10 proportion)
    > Check for the feasibility of dimensionality reduction so that only relevant attributes gets into the model which will further help in efficiency and optimization of the model
    
    > Try to build an staged model which will filter Supercategory -> Subcategory -> Maincategory at first level, Color/Hex Code at Second Level and further attributes at final level

        > SubCategory Classification --> Done

        > Color Classification 

        > Attributes Classification

+ Train the model using two three major classifiers like SVC, Random Forest Classifier, tree based models like LightGBM, XGBoost and Neural Network etc.
+ Setup the MobileNet_V2 with weights = imagenet for the feature extraction from the input image

        > Status: Done
        
+ Get the output vector basis the model and flatten the same for the validation
+ Test for the accuracy, efficiency and optimization using advanced techniques and required metrics
+ Bundle the whole model and create an MVP using Streamlit Application for the realtime testing purpose